### ISSUE TITLE
"hugo server -w" ignores config.toml changes after REMOVE event since 0.38

### ISSUE THREAD

**Issue by anthonyfok** [gohugoio/hugo#4701]
... and after 1 save in gedit (GNOME Text Editor), as tested with Hugo 0.40.2 on Debian sid (buster, testing) on amd64.

When I try to save config.toml repeatedly every few seconds:

* For Vim, Hugo receives a RENAME event the first save (and rebuilds), and CHMOD+REMOVE or just REMOVE event the second save (and rebuilds), and after that, Vim seems to forget about config.toml altogether, and does not receive further events from changes to config.toml, unless I press Ctrl+C and restart `hugo server`.

* For gedit, Hugo receives CHMOD+REMOVE for the first save (and rebuilds), and ignores config.toml changes from then on unless `hugo server` is restarted.

Maybe it has to do with how currently Hugo handles the REMOVE event?  Or something else altogether?  (I haven't checked the code yet at all.)

Thanks!

**Comment by anthonyfok** [comment]
It turns out that hugo watches the file `config.toml` (or .yaml) itself, instead of watching the directory that contains `config.yaml`.  (Or does it?)

Turns out it was a previously solved issue back in 2015/2016,, see:

* spf13/viper#142: WatchConfig() does not work reliably with certain text editors
* See also fsnotify/fsnotify#80 and fsnotify/fsnotify#94

**Comment by anthonyfok** [comment]
This issue appears to be a regression that started in Hugo 0.38 (April 2018), up to the latest 0.40.2 as of today.

Hugo 0.16 (June 2016) to 0.37.1 (March 2018) were good.

**Comment by bep** [comment]
It used to watch the directory, but that really does not work (the root contains all sort of stuff).

**Comment by anthonyfok** [comment]
> It used to watch the directory, but that really does not work (the root contains all sort of stuff).

Oh, I see, thanks for the feedback!

I didn't look at the code carefully, but at first glance it seems that Hugo is now watching config.toml as a file instead of relying Viper's WatchConfig() to do the job, but I wasn't sure.  Thank you for your confirmation!

I was thinking perhaps there are workarounds like re-adding config.toml to the watch list after a REMOVE event.  I will experiment.  :-)

**Comment by jpouellet** [comment]
#4720 does not seem to have been sufficient to resolve this.

With vim and both current hugo master (10ac2ec) and latest release (v0.49/extended) I am able to reproduce the problem of only the first change to config.toml causing a rebuild.

`hugo serve` output:
```
/hugo serve

                   | EN  
+------------------+----+
  Pages            |  7  
  Paginator pages  |  0  
  Non-page files   |  0  
  Static files     |  3  
  Processed images |  0  
  Aliases          |  0  
  Sitemaps         |  1  
  Cleaned          |  0  

Total in 13 ms
Watching for changes in /home/user/quickstart/{content,data,layouts,static,themes}
Watching for config changes in /home/user/quickstart/config.toml
Serving pages from memory
Running in Fast Render Mode. For full rebuilds on change: hugo server --disableFastRender
Web Server is available at http://localhost:1313/ (bind address 127.0.0.1)
Press Ctrl+C to stop

Change detected, rebuilding site
2018-10-03 00:43:36.485 -0400
Total in 4 ms
```
(and there is no further output, regardless of how many times I further edit config.toml)

---

Output of program in https://github.com/howeyc/fsnotify/issues/68#issue-20804151 (linked to from https://github.com/gohugoio/hugo/issues/3850) while changing someDir/foo is:

open vim:
```
2018/10/03 00:48:27 notify.go:23: event: "testDir/.foo.swp": CREATE
2018/10/03 00:48:27 notify.go:23: event: "testDir/.foo.swpx": DELETE
2018/10/03 00:48:27 notify.go:23: event: "testDir/.foo.swp": DELETE
2018/10/03 00:48:27 notify.go:23: event: "testDir/.foo.swp": CREATE
2018/10/03 00:48:27 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:27 notify.go:23: event: "testDir/.foo.swp": MODIFY|ATTRIB
2018/10/03 00:48:31 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:33 notify.go:23: event: "testDir/4913": CREATE
2018/10/03 00:48:33 notify.go:23: event: "testDir/4913": DELETE
2018/10/03 00:48:33 notify.go:23: event: "testDir/foo": RENAME
2018/10/03 00:48:33 notify.go:23: event: "testDir/foo~": CREATE
2018/10/03 00:48:33 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:33 notify.go:23: event: "testDir/foo": CREATE
2018/10/03 00:48:33 notify.go:23: event: "testDir/foo": MODIFY
2018/10/03 00:48:33 notify.go:23: event: "testDir/foo": MODIFY|ATTRIB
2018/10/03 00:48:33 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:33 notify.go:23: event: "testDir/foo~": DELETE
2018/10/03 00:48:37 notify.go:23: event: "testDir/.foo.swp": MODIFY
```

edit file:
```
2018/10/03 00:48:40 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:44 notify.go:23: event: "testDir/4913": DELETE
2018/10/03 00:48:44 notify.go:23: event: "testDir/foo": RENAME
2018/10/03 00:48:44 notify.go:23: event: "testDir/foo~": CREATE
2018/10/03 00:48:44 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:44 notify.go:23: event: "testDir/foo": CREATE
2018/10/03 00:48:44 notify.go:23: event: "testDir/foo": MODIFY
2018/10/03 00:48:44 notify.go:23: event: "testDir/foo": MODIFY|ATTRIB
2018/10/03 00:48:44 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:44 notify.go:23: event: "testDir/foo~": DELETE
```

edit file again:
```
2018/10/03 00:48:48 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:49 notify.go:23: event: "testDir/4913": DELETE
2018/10/03 00:48:49 notify.go:23: event: "testDir/foo": RENAME
2018/10/03 00:48:49 notify.go:23: event: "testDir/foo~": CREATE
2018/10/03 00:48:49 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:49 notify.go:23: event: "testDir/foo": CREATE
2018/10/03 00:48:49 notify.go:23: event: "testDir/foo": MODIFY
2018/10/03 00:48:49 notify.go:23: event: "testDir/foo": MODIFY|ATTRIB
2018/10/03 00:48:49 notify.go:23: event: "testDir/.foo.swp": MODIFY
2018/10/03 00:48:49 notify.go:23: event: "testDir/foo~": DELETE
2018/10/03 00:48:53 notify.go:23: event: "testDir/.foo.swp": MODIFY
```

close vim:
```
2018/10/03 00:48:55 notify.go:23: event: "testDir/.foo.swp": DELETE
```

**Comment by anthonyfok** [comment]
@jpouellet Thanks for letting me know.  Strange, I was certain that the fix #4720 used to work for me, but no longer when I recompiled it.  And heck, the Hugo v0.47 release binary, the first released version to contain this fix, no longer works for me either.  Strange...

Anyhow, since the same problem has been reported again in #5205, let's continue there.  I'll keep this old issue closed.
